servant-client-ghcjs/src/Servant/Client/Internal/XhrClient.hs

```diff
@@ -0,0 +1,241 @@
+import           Control.Monad.Trans.Except
+import qualified Data.ByteString.Char8 as BS
+import           Data.CaseInsensitive
+import           Data.Char
+import           Data.Foldable               (toList)
+import           Data.Functor.Alt            (Alt (..))
+import           Data.Proxy                  (Proxy (..))
+import qualified Data.Sequence           as Seq
+import           Data.String.Conversions
+import           Foreign.StablePtr
+import           GHC.Generics
+import           GHCJS.Foreign.Callback
+import           GHCJS.Prim
+import           Network.HTTP.Types
+import           Servant.Client.Core
+
+newtype JSXMLHttpRequest = JSXMLHttpRequest JSVal
+
+newtype JSXMLHttpRequestClass = JSXMLHttpRequestClass JSVal
+
+newtype GhcjsClientEnv
+   = GhcjsClientEnv
+   { baseUrl :: BaseUrl }
+   deriving (Eq, Show)
+
+client :: HasClient GhcjsClientM api => Proxy api -> Client GhcjsClientM api
+client api = api `clientIn` (Proxy :: Proxy GhcjsClientM)
+
+newtype GhcjsClientM a = GhcjsClientM
+  { runGhcjsClientM' :: ReaderT GhcjsClientEnv (ExceptT ServantError IO) a }
+  deriving ( Functor, Applicative, Monad, MonadIO, Generic
+           , MonadReader GhcjsClientEnv, MonadError ServantError, MonadThrow
+           , MonadCatch)
+
+instance MonadBase IO GhcjsClientM where
+  liftBase = GhcjsClientM . liftBase
+
+instance MonadBaseControl IO GhcjsClientM where
+  type StM GhcjsClientM a = Either ServantError a
+
+  liftBaseWith f = GhcjsClientM (liftBaseWith (\g -> f (g . runGhcjsClientM')))
+
+  restoreM st = GhcjsClientM (restoreM st)
+
+-- | Try clients in order, last error is preserved.
+instance Alt GhcjsClientM where
+  a <!> b = a `catchError` const b
+
+instance RunClient GhcjsClientM where
+  runRequest = performRequest
+  throwServantError = throwError
+  catchServantError = catchError
+
+instance ClientLike (GhcjsClientM a) (GhcjsClientM a) where
+  mkClient = id
+
+runGhcjsClientM :: GhcjsClientM a -> GhcjsClientEnv -> IO (Either ServantError a)
+runGhcjsClientM cm env = runExceptT $ flip runReaderT env $ runGhcjsClientM' cm
+
+performRequest :: Request -> GhcjsClientM Response
+performRequest req = do
+  xhr <- liftIO initXhr
+  burl <- asks baseUrl
+  liftIO $ performXhr xhr burl req
+  toResponse xhr
+
+-- * initialization
+
+initXhr :: IO JSXMLHttpRequest
+initXhr = do
+  lib <- requireXMLHttpRequestClass
+  newXMLHttpRequest lib
+
+foreign import javascript unsafe
+  -- branching between node (for testing) and browsers
+  "(function () {if (typeof require !== 'undefined') { return require('xhr2'); } else { return XMLHttpRequest; };})()"
+  requireXMLHttpRequestClass :: IO JSXMLHttpRequestClass
+
+foreign import javascript unsafe "new $1()"
+  newXMLHttpRequest :: JSXMLHttpRequestClass -> IO JSXMLHttpRequest
+
+-- * performing requests
+-- Performs the xhr and blocks until the response was received
+performXhr :: JSXMLHttpRequest -> BaseUrl -> Request -> IO ()
```

Comment:
A lot of the FFI calls and `xhr` functionality seems to be present in [ghcjs-base](https://github.com/ghcjs/ghcjs-base/blob/master/JavaScript/Web/XMLHttpRequest.hs). Is there a reason not to use that?

Reply:
It's possible, but status text would be lost, as its API doesn't support getting the status text. Besides there would have to be some conversion between servant and ghcjs' types (Request, Response, the exceptions). Using ffi is somewhat more direct.